accounts/abi: allow abi: tags when unpacking structs

adriamb opened the pull request:
Allow to unpack structures using an `abi` tag instead the variable name, ala golang with json fields, e.g.

```
type MyEventParameters struct {
     Value1 *big.Int `abi:"value"`
}
```

holiman commented:
The code looks pretty good, but it would be good with additional testcases. For example, 

- What happens when a struct has multiple fields, some tagged and some not tagged?
  - e.g. a struct several `uint` called `value1`, `value2` etc, and the last one `value5` is tagged `value1`? Does it overwrite the first item in the struct, or what happens? 

Try to think up some failure-cases, and encode them in the form of tests. Tests should not only verify the 'happy-path', but also note the behaviour at edge-cases and error-cases.

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: da950241870886ce2af4afc47409ce1e5e1873ae
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

adriamb commented:
Hi @holiman, added a decent check about unpack struct sanity and some tests, I think I am covering the enough the test cases.

BTW, forgot to prefix the last commit with the package name, there is any way to fix it?

Thanks!

holiman commented:
Great @adriamb , I just looked through it, thanks! 

Regarding the commit: you can rebase it

```
git rebase -i HEAD~2
// squash the second one into the first, or just change the commit message
git push -f 
```

holiman commented on accounts/abi/reflect.go:
```diff
@@ -111,8 +112,8 @@ func requireUnpackKind(v reflect.Value, t reflect.Type, k reflect.Kind,
 
-// requireUniqueStructFieldNames makes sure field names don't collide
-func requireUniqueStructFieldNames(args Arguments) error {
+// requireUniqueStructFieldNames makes sure abi field names don't collide
```
s/requireUniqueStructFieldNames/requireUniqueAbiFieldNames

holiman commented on accounts/abi/argument.go:
```diff
@@ -113,7 +113,10 @@ func (arguments Arguments) unpackTuple(v interface{}, marshalledValues []interfa
+		if err := requireUniqueAbiFieldNames(arguments); err != nil {
+			return err
+		}
+		if err := requireUniqueStructFieldNames(typ); err != nil {
```
How come that when unpacking a tuple, you check both for `requireUniqueAbiFieldNames` and `requireUniqueStructFieldNames`, but not when unpacking atomic?

adriamb replied:
Mmm... you're right. I also will add a test to cover this case.

holiman commented on accounts/abi/reflect.go:
```diff
@@ -126,3 +127,36 @@ func requireUniqueStructFieldNames(args Arguments) error {
+			}
+			exists[tagName] = true
+		} else {
+			fieldAsTag := strings.ToLower(fieldName[:1]) + fieldName[1:]
```
What does `ToLower` do here? How come you don't just use `fieldName`?

adriamb replied:
Oh, thanks @holiman, `abi:` tags, allows more corner cases that I assumed at the beginning, like allowing events like `event DoIt(uint a, uint A);` or  `event DoIt(uint a, uint);`... I did not realize about it.

I need to be compatible with the previous logic without `abi:` and the new one. I need to work more on this.

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: da950241870886ce2af4afc47409ce1e5e1873ae
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com

adriamb commented:
@holiman, changed the implementation, now:

1) Added a `mapAbiToStructFields` function that computes how ABI fields are mapped into a struct. Now `requireUniqueStructFieldNames` and `unpackStruct` are no longer necessary, and IMO makes cleaner.
2) Is compatible with the previous mapping logic (passes all existing tests without modifying).
3) Added more tests to check the functionality.

I did't added functionality to handle unnamed abi variables for now. Do you think that makes sense to add this functionality? If yes, what about allowing to reference the abi fiends by their position?

`abi: [ { name:"" , name: "" } ]`

```
type S struct {
    Field1 *big.Int `abi:"[0]"` // field at position 0
    Field2 *big.Int `abi:"[1]"`// field at position 1
}
```

Thanks!

holiman commented:
Strangely, the travis message on the ticket reports as failed/cancelled, but it looks fine here: https://travis-ci.org/ethereum/go-ethereum/builds/375286042 

@adriamb regarding unnamed fields, right now those would be filled in the order written, right? If that's the case, I don't think there's anything to gain by having yet another way to reference them. It will just add more quirky edgecases. I don't have a strong opinion here.

holiman commented:
Ping @gballet : please look through and merge or commentOn May 10, 2018 10:41 AM, aka codecontext <notifications@github.com> wrote:@holiman, any other task to do this improvement?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

gballet commented on accounts/abi/reflect.go:
```diff
@@ -111,18 +112,99 @@ func requireUnpackKind(v reflect.Value, t reflect.Type, k reflect.Kind,
-		if exists[field] {
-			return fmt.Errorf("abi: multiple outputs mapping to the same struct field '%s'", field)
+
+		// this abi been already paired, skip
```
this abi has already been paired, skip...

gballet commented on accounts/abi/reflect.go:
```diff
@@ -111,18 +112,99 @@ func requireUnpackKind(v reflect.Value, t reflect.Type, k reflect.Kind,
+
+		// this abi been already paired, skip
+		if abi2struct[abiFieldName] != "" {
+			// unless if exists another not still assigned struct field with the same name
```
unless there is another yet unassigned struct field with the same name

I think you mean mean "the same "abi" name", right?

adriamb replied:
Ok, I will add a clarification there.

gballet commented on accounts/abi/reflect.go:
```diff
@@ -111,18 +112,99 @@ func requireUnpackKind(v reflect.Value, t reflect.Type, k reflect.Kind,
-		exists[field] = true
+
+		// error if this struct field been already paired.
+		if struct2abi[structFieldName] != "" {
```
could you move this before line 181? It would simplify the condition check on line 184

gballet commented on accounts/abi/reflect.go:
```diff
@@ -111,18 +112,99 @@ func requireUnpackKind(v reflect.Value, t reflect.Type, k reflect.Kind,
 		}
-		exists[field] = true
+
+		// error if this struct field been already paired.
```
return an error if this struct field has already been paired

gballet commented on accounts/abi/reflect.go:
```diff
@@ -111,18 +112,99 @@ func requireUnpackKind(v reflect.Value, t reflect.Type, k reflect.Kind,
+			continue
+		}
+
+		// check if tag is emmpty.
```
empty

gballet commented on accounts/abi/reflect.go:
```diff
@@ -111,18 +112,99 @@ func requireUnpackKind(v reflect.Value, t reflect.Type, k reflect.Kind,
+// first round: for each Exportable field that contains a `abi:""` tag
+//   and this field name exists in the arguments, pair them together.
+// second round: for each argument field that has not been already linked,
+//   find what variable is expected to be mapped into, if exists and has not been
```
if **it** exists

gballet commented on accounts/abi/reflect.go:
```diff
@@ -111,18 +112,101 @@ func requireUnpackKind(v reflect.Value, t reflect.Type, k reflect.Kind,
-		if exists[field] {
-			return fmt.Errorf("abi: multiple outputs mapping to the same struct field '%s'", field)
+
+		// this abi has already been paired, skip... unless exists another not still assigned
```
apparently I missed this one:
unless there exists another, yet unassigned...

adriamb replied:
Haha, thanks for the grammar checks @gballet , I'll change it! [ and I annotate to use a grammar checker next time for the comments :) ]

I have no strong opinion about the default value. No problem about implementing it. In which use case are you thinking about? A generic event / function argument parser?

gballet replied:
I was thinking about default values for API calls, but it doesn't really make sense actually because that would be a Go-only behavior. Just fix the last grammar thing. @holiman and I will discuss the default value topic, but I don't think it's necessary after all.

holiman replied:
> discuss the default value topic, but I don't think it's necessary after all.

I agree

gballet replied:
indeed so @adriamb please fix the last grammar issue and we'll merge your PR.

gballet replied:
nvm, I just did

GitCop commented:
Thank you for your contribution! Your commits seem to not adhere to the repository coding standards

 * Commit: cae6fc3349c1767dc44c5c0248951d3e355f7718
  * Commits must be prefixed with the package(s) they modify

Please check the [contribution guidelines](https://github.com/ethereum/go-ethereum#contribution) for more details.

---
This message was auto-generated by https://gitcop.com
